prov/shm/src/smr_ep.c

```diff
@@ -0,0 +1,747 @@
+	cmd->hdr.op.flags = 0;
+
+	cmd->hdr.op.data = 0;
+	cmd->hdr.op.resv = 0;
+
+	cmd->hdr.op.size = ofi_copy_from_iov(cmd->data.msg, SMR_CMD_DATA_LEN,
+					     msg->msg_iov, msg->iov_count, 0);
+}
+
+static void smr_format_inject(struct smr_cmd *cmd, const struct fi_msg *msg,
+			      struct smr_region *smr, struct smr_inject_buf *tx_buf)
+{
+	cmd->hdr.op.version = OFI_OP_VERSION;
+	cmd->hdr.op.rx_index = 0;
+	cmd->hdr.op.op = ofi_op_msg;
+	cmd->hdr.op.op_data = smr_op_inject;
+	cmd->hdr.op.flags = 0;
+
+	cmd->hdr.op.size = 0;
+	cmd->hdr.op.resv = 0;
+	cmd->hdr.op.data = (uint64_t) ((char **) tx_buf - (char **) smr);
+
+	cmd->hdr.op.size = ofi_copy_from_iov(tx_buf->data, SMR_INJECT_SIZE,
+					     msg->msg_iov, msg->iov_count, 0);
+}
+
+static void smr_format_cma(struct smr_cmd *cmd, const struct fi_msg *msg,
+			   size_t total_len, struct smr_region *smr,
+			   struct smr_resp *resp)
+{
+	int i;
+
+	cmd->hdr.op.version = OFI_OP_VERSION;
+	cmd->hdr.op.rx_index = 0;
+	cmd->hdr.op.op = ofi_op_msg;
+	cmd->hdr.op.op_data = smr_op_iov;
+	cmd->hdr.op.flags = 0;
+
+	cmd->hdr.op.size = 0;
+	cmd->hdr.op.resv = 0;
+	resp->status = -FI_EBUSY;
+	cmd->hdr.op.data = (uint64_t) ((char **) resp - (char **) smr);
+	cmd->hdr.op.iov_count = msg->iov_count;
+
+	for (i = 0; i < msg->iov_count; i++) {
+		cmd->data.iov[i].iov_base = msg->msg_iov[i].iov_base;
+		cmd->data.iov[i].iov_len = msg->msg_iov[i].iov_len;
+		cmd->hdr.op.size += msg->msg_iov[i].iov_len;
+	}
+}
+
+ssize_t smr_sendmsg(struct fid_ep *ep_fid, const struct fi_msg *msg,
+		uint64_t flags)
+{
+	struct smr_ep *ep;
+	struct smr_region *peer_smr;
+	struct smr_inject_buf *tx_buf;
+	struct smr_resp *resp;
+	struct smr_cmd *cmd;
+	int peer_id;
+	ssize_t ret = 0;
+	size_t total_len;
+
+	if (msg->iov_count > SMR_IOV_LIMIT) {
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+			"Too many iovs\n");
+		return -FI_EINVAL;
+	}
+
+	ep = container_of(ep_fid, struct smr_ep, util_ep.ep_fid.fid);
+	peer_id = *(int *)ofi_av_get_addr(ep->util_ep.av, msg->addr);
+	fastlock_acquire(&ep->util_ep.tx_cq->cq_lock);
+	if (ofi_cirque_isfull(ep->util_ep.tx_cq->cirq)) {
+		ret = -FI_EAGAIN;
+		goto out;
+	}
+
+	peer_smr = smr_peer_region(ep->region, peer_id);
+	if (ep->region->map->peer_addr[peer_id].name[0] != '\0') {
+		ret = smr_map_to_region(&smr_prov, &peer_smr,
+					ep->region->map->peer_addr[peer_id].name);
+		if (ret) {
+			if (ret == -ENOENT)
+				ret = -FI_EAGAIN; 
+			goto out;
+		}
+	}
+
+	fastlock_acquire(&peer_smr->lock);
+	if (ofi_cirque_isfull(smr_cmd_queue(peer_smr))) {
+		ret = -FI_EAGAIN;
+		goto unlock;
+	}
+
+	total_len = ofi_total_iov_len(msg->msg_iov, msg->iov_count);
+
+	cmd = ofi_cirque_tail(smr_cmd_queue(peer_smr));
+
+	if (total_len <= SMR_CMD_DATA_LEN) {
+		smr_format_send(cmd, msg);
+	} else if (total_len <= SMR_INJECT_SIZE) {
+		freestack_pop(smr_inject_pool(peer_smr), tx_buf);
+		smr_format_inject(cmd, msg, peer_smr, tx_buf);
+	} else {
+		if (ofi_cirque_isfull(smr_resp_queue(ep->region))) {
```

Comment:
We may be able to assert() this.  If there's space in the tx cq, there should be space available in the resp_queue.  Please walk though the cq/resp queue and see if this is the case.  It seems like it would naturally fall out this way.  (The tx cq should be the limiting factor, not the resp queue).